stablehlo/transforms/VhloToVersion.cpp

```diff
@@ -296,15 +297,146 @@ struct VersionConversionPattern : OpConversionPattern<SourceOp> {
 /// Upgrade and Downgrade Definitions ///
 /////////////////////////////////////////
 
+StringV1Attr convert(SymbolRefV1Attr attr) {
+  return StringV1Attr::get(attr.getContext(), attr.getValue());
+}
+SymbolRefV1Attr convert(StringV1Attr attr) {
+  return SymbolRefV1Attr::get(attr.getContext(), attr.getValue());
+}
+
+template <typename OpType, typename SourceAttr, typename TargetAttr>
+struct AttributeConversionPattern : public OpConversionPattern<OpType> {
+  using OpConversionPattern<OpType>::OpConversionPattern;
+  virtual SmallVector<StringRef> getAttributeNames(OpType op) const = 0;
+
+  FailureOr<SmallVector<SourceAttr>> matchArrayOfSource(Attribute attr) const {
+    vhlo::ArrayV1Attr arrayAttr = dyn_cast<ArrayV1Attr>(attr);
+    if (!arrayAttr) return failure();
+    SmallVector<SourceAttr> elements;
+    for (Attribute attr : arrayAttr.getValue()) {
+      SourceAttr element = dyn_cast<SourceAttr>(attr);
+      if (!element) return failure();
+      elements.push_back(element);
+    }
+    return elements;
+  }
+
+  LogicalResult convertAttribute(ConversionPatternRewriter& rewriter, OpType op,
+                                 StringRef attrName) const {
+    Attribute attr = op->getAttr(attrName);
+
+    // Support the following:
+    //   Array<SourceAttr> -> Array<TargetAttr>
+    //   SourceAttr -> TargetAttr
+    if (auto array = matchArrayOfSource(attr); succeeded(array)) {
```

Comment:
This is fine, but I feel like there should be some built-in way to tell MLIR that this attribute contains other attributes, and have MLIR do the visiting as needed? I don't know of any such facility though.